lib.customisation.makeScope: Make `overrideScope` consistent with `makeScopeWithSplicing`

Artturin opened the pull request:
Right now converting `makeScope` to `makeScopeWithSplicing` is not transparent to users and requires adding a warning for `overrideScope'` in the set itself.

Warning and `overrideScope'` were added in 2018 b9dce11712d2bfc8cd367df5a7f737a5cec1e252 and there should be no users left after 5 years.

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Artturin committed:
lib.customisation.makeScope: Make `overrideScope` consistent with `makeScopeWithSplicing`

Right now converting `makeScope` to `makeScopeWithSplicing` is not
transparent to users and requires adding a warning for `overrideScope'`
in the set itself.

Warning and `overrideScope'` were added in 2018 b9dce11712d2bfc8cd367df5a7f737a5cec1e252 and there should be no users left after 5 years.

infinisil commented on lib/customisation.nix:
```diff
@@ -269,9 +269,8 @@ rec {
-            "`overrideScope` (from `lib.makeScope`) is deprecated. Do `overrideScope' (self: super: { … })` instead of `overrideScope (super: self: { … })`. All other overrides have the parameters in that order, including other definitions of `overrideScope`. This was the only definition violating the pattern."
-            (makeScope newScope (lib.fixedPoints.extends (lib.flip g) f));
+          overrideScope = g: makeScope newScope (lib.fixedPoints.extends g f);
+          # Add a warning after 23.11 and remove after 24.05.
```
I think the timing should be much more relaxed. Somebody could be using `overrideScope` while still using 23.05 and get the warning that it's deprecated and `overrideScope'` should be used, but then they upgrade to 24.05 on the same day and suddenly `.overrideScope'` is gone, even though they only upgraded by 2 release.

Artturin replied:
I meant remove it after 24.05 is branched off

Artturin replied:
I'll tweak these, push coming soon.

amjoseph-nixpkgs replied:
Agree.  Looks like this was resolved in the latest push.

infinisil replied:
Not really: https://github.com/NixOS/nixpkgs/pull/248988#discussion_r1293613666

infinisil reviewed (changes requested):
Oof, the fact that `overrideScope'` didn't even exist on `makeScopeWithSplicing` is pretty bad, because depending on the set, only one of `overrideScope` or `overrideScope'` would work, and the other would warn or error, without a safe choice that always worked.

I'm generally okay with this change though

Artturin committed:
treewide: `overrideScope'` -> `overrideScope`

`lib.makeScope` `overrideScope'` has been renamed to `overrideScope`

`fd --type f | xargs sd --string-mode "overrideScope'" "overrideScope"`

Artturin commented on lib/customisation.nix:
@infinisil 

My plan now is to backport the first commit to 23.05 so users can support stable and unstable at the same time

amjoseph-nixpkgs reviewed (approved):
LGTM.  Looks like @infinisil's [concern](https://github.com/NixOS/nixpkgs/pull/248988#pullrequestreview-1576013928) was addressed in the latest push.

infinisil commented on lib/customisation.nix:
```diff
@@ -269,10 +269,11 @@ rec {
-          overrideScope' = g: makeScope newScope (lib.fixedPoints.extends g f);
+          overrideScope = g: makeScope newScope (lib.fixedPoints.extends g f);
+          # Remove after 24.05 is released.
+          overrideScope' = g: lib.warn
+            "`overrideScope'` (from `lib.makeScope`) has been renamed to `overrideScope`."
```
I needs to use `warnIf (isInOldestRelease 2311)`, such that we only warn once 23.11 isn't supported anymore, otherwise we still have the problem from https://github.com/NixOS/nixpkgs/pull/248988#discussion_r1292940144, because there wouldn't be a transition period where users could use both versions of the function.

Artturin replied:
https://github.com/NixOS/nixpkgs/pull/248988#discussion_r1292984930

> My plan now is to backport the first commit to 23.05 so users can support stable and unstable at the same time

infinisil replied:
I don't think that's a great idea, because then you can't just check whether a release is 23.05 to figure out if `overrideScope` is fine to use, instead you have to make sure it's 23.05 after a specific commit/date.

Using `warnIf (isInOldestRelease ..)` without a backport would be my preference, it means 23.11 will be a clear boundary of when it changed.

And anyways, there's no need to hurry with this deprecation, let's take it slow

Artturin replied:
Uses warnIf now

Artturin commented on lib/customisation.nix:
```diff
@@ -269,10 +269,11 @@ rec {
-            (makeScope newScope (lib.fixedPoints.extends (lib.flip g) f));
-          overrideScope' = g: makeScope newScope (lib.fixedPoints.extends g f);
+          overrideScope = g: makeScope newScope (lib.fixedPoints.extends g f);
+          # Remove after 24.05 is released.
```
23.11 = warn
24.05 = warn
24.11 = removed

so 2 releases of warnings

infinisil commented:
Please don't self-merge when the PR is actively being reviewed and commented on..

Artturin commented:
> Please don't self-merge when the PR is actively being reviewed and commented on..
> 
> (My approval happened at the exact same second as the merge, did you enable auto-merge?)

Yes enabled auto merge

infinisil commented:
I see, I think that's okay then, though I still think self-merges should be avoided when possible (in this case I would've merged it for you)

Artturin commented:
Thanks for the reviews

Artturin commented:
Yeah, this was the first PR I saw it in and decided to try it.
